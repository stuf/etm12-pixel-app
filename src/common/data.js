/**
 * @module data
 * @namespace common
 *
 * Common curried data structure constructors
 */
import * as R from 'ramda';

export const mkClampedArray = R.constructN(1, Uint8ClampedArray);

export const mkImageData = R.constructN(3, ImageData);
