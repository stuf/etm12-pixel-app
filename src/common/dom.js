/**
 * @param {string} type
 * @param {object} attrs
 * @return {HTMLElement}
 */
export const mkElem = (type, attrs) =>
  Object.assign(document.createElement(type), attrs);
