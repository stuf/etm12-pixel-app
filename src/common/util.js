import * as R from 'kefir.ramda';

import { LiftedAry1Fn } from 'types.d';

/**
 *
 * @param {(string|number)} b
 * @param {(string|number)} a
 * @return {string}
 */
export const withSuffix = (b, a) => String.prototype.concat(a, b);

export const toUnit = unit => withSuffix.bind(null, unit);

export const asPercent = toUnit('%');
export const asRem = toUnit('rem');
export const asPx = toUnit('px');
export const asVw = toUnit('vw');
export const asVh = toUnit('vh');

//

/**
 * @type {LiftedAry1Fn<number, number>}
 */
export const reciprocal = R.divide(1);
