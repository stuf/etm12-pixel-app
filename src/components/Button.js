import * as React from 'karet';

import * as T from './Button';

import style from './Button.module.scss';

/**
 * @param {T.Props} props
 */
function Button({ children }) {
  return <button className={style.root}>{children}</button>;
}

export default Button;
