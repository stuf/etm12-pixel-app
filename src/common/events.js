/* eslint no-unused-vars: [1, {"varsIgnorePattern": "(LiftedAry[1-3]Fn|Size|K)"}] */
/**
 * @module events
 * @namespace common
 */
import * as U from 'karet.util';
import * as R from 'kefir.ramda';
import * as K from 'kefir';

/**
 *
 * @param {string} type
 * @param {HTMLElement} source
 * @return {K.Property<Event, any>}
 */
export const takeEvents = (type, source) =>
  U.thru(
    source,
    U.flatMapLatest(src => U.fromEvents(src, type, a => a)),
    U.toProperty,
  );

//

export const pagePos = R.props(['pageX', 'pageY']);
export const screenPos = R.props(['screenX', 'screenY']);
export const layerPos = R.props(['layerX', 'layerY']);
