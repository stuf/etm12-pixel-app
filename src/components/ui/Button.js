/* eslint no-unused-vars: [1, {"varsIgnorePattern": "[K|T|Model]"}] */
import * as React from 'karet';
import * as U from 'karet.util';

function Button({ id, children, disabled, action, group }) {
  return (
    <button
      {...{
        id,
        className: U.cns('button', U.when(group, '--group')),
        disabled,
        onClick: U.actions(U.preventDefault, action),
      }}
    >
      {children}
    </button>
  );
}

export default Button;
