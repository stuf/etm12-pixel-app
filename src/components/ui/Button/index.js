/* eslint no-unused-vars: [1, {"varsIgnorePattern": "[K|T|Model]"}] */
import * as React from 'karet';
import * as U from 'karet.util';

import * as T from './index.d';
import style from './index.module.scss';

/**
 * @param {T.Props} props
 * @return {T.Component}
 */
function Button({ children, action }) {
  return (
    <button
      className={style.root}
      onClick={U.actions(U.preventDefault, action)}
    >
      {children}
    </button>
  );
}

export default Button;
