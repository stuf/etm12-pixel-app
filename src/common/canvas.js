/* eslint no-unused-vars: [1, {"varsIgnorePattern": "(LiftedAry[1-3]Fn|Size)"}] */
import * as U from 'karet.util';
import * as K from 'kefir';
import { saveAs } from 'file-saver';
import Long from 'long';

import { takeEvents } from 'common/events';
import { mkElem } from 'common/dom';
import { mkClampedArray, mkImageData } from 'common/data';

import { LiftedAry1Fn, LiftedAry2Fn, Size } from 'types.d';

const COLOR_CHANNELS = 4;

/**
 * @type {LiftedAry2Fn<Size, number, Size>}
 */
export const scaleSize = U.lift(([w, h], m) => [w * m, h * m]);

/**
 * @type {LiftedAry1Fn<HTMLCanvasElement, CanvasRenderingContext2D>}
 */
export const getContext = U.lift(el => el.getContext('2d'));

/**
 * @type {LiftedAry2Fn<CanvasRenderingContext2D, Size, ImageData>}
 */
export const getImageData = U.lift((ctx, [w, h]) =>
  ctx.getImageData(0, 0, w, h),
);

/**
 * @type {LiftedAry3Fn<CanvasRenderingContext2D, number, number, ImageData>}
 */
export const getImageData_ = U.lift((ctx, w, h) =>
  ctx.getImageData(0, 0, w, h),
);

//

/**
 * @param {HTMLElement} domRef
 */
export const drawingEvents = domRef => {
  const drag = takeEvents('mousedown', domRef)
    .flatMapLatest(() =>
      takeEvents('mousemove', domRef).takeUntilBy(
        K.merge([
          takeEvents('mouseup', domRef),
          takeEvents('mouseout', domRef),
        ]).take(1),
      ),
    )
    .toProperty();

  const click = takeEvents('mousedown', domRef);

  return K.merge([drag, click]).toProperty();
};

//

/**
 * @type {LiftedAry2Fn<[number, number], number, number]>}
 */
export const getIx = U.lift(
  ([x, y], w) => (Math.trunc(y) * w + Math.trunc(x)) * COLOR_CHANNELS,
);

/**
 * @type {LiftedAry2Fn<[number, number], number, number]>}
 */
export const getIxRange = U.lift(([x, y], w) => {
  const ix = getIx([x, y], w);
  return [ix, ix + COLOR_CHANNELS];
});

// Colors

const _hexFromString = str => Long.fromString(str, null, 16);

/**
 * Convert a string representing a hexadecimal color, optionally
 * including an alpha value, into a 4-tuple (RGBA).
 */
export const convertFromHexColor = U.lift(str => {
  if (str.length === 6) {
    const val = parseInt(str, 16);
    return [(val & 0xff0000) >> 16, (val & 0xff00) >> 8, val & 0xff, 255];
  }

  const val = Long.fromString(str, null, 16);
  const rmask = _hexFromString('ff000000');
  const gmask = _hexFromString('ff0000');
  const bmask = _hexFromString('ff00');
  const amask = _hexFromString('ff');

  return [
    +val.and(rmask).shr(24),
    +val.and(bmask).shr(16),
    +val.and(gmask).shr(8),
    +val.and(amask),
  ];
});

/**
 * @type {LiftedAry1Fn<[number, number], number[]>}
 */
export const empty = U.lift(([w, h]) => Array(w * h * COLOR_CHANNELS).fill(0));

//

/**
 *
 * @param {number[]} image
 * @param {[number, number]} size
 * @param {string} filename
 */
export const saveImage = U.lift((data, size, filename) => {
  const [width, height] = size;

  const el = mkElem('canvas', { width, height });

  const ctx = getContext(el);

  const clampedData = mkClampedArray(data);
  const imageData = mkImageData(clampedData, width, height);

  ctx.putImageData(imageData, 0, 0);
  el.toBlob(b => saveAs(b, filename));
});
