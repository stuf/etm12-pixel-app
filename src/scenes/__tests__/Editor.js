import * as React from 'karet';
import * as U from 'karet.util';
import * as H from 'kefir.partial.lenses.history';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import EditorScene from 'scenes/Editor';

import { empty } from 'common/canvas';

import Button from 'components/ui/Button';

it('matches snapshot', () => {
  const size = [4, 4];

  const state = U.atom({
    canvas: { size, scale: 16 },
    color: { currentPalette: 0, currentColor: 0, palettes: [] },
  });

  const canvasData = U.atom(
    H.init({ maxCount: 10, replacePeriod: 200 }, empty(size)),
  );

  const w = mount(<EditorScene {...{ state, canvasData }} />);

  const w2 = mount(
    <EditorScene {...{ state, canvasData: canvasData.get() }} />,
  );

  const saveImageButton = w.find('#save-current-image').first();
  saveImageButton.simulate('click');

  state.view('size').set([16, 16]);
});
