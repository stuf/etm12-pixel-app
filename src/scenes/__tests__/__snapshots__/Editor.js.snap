// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<EditorScene
  canvasData={
    Object {
      "c": Object {
        "e": true,
        "m": 9,
        "p": 200,
      },
      "i": 0,
      "t": Object {
        "l": 0,
        "r": Array [
          1572278123270,
        ],
        "u": 1,
      },
      "v": Object {
        "l": 0,
        "r": Array [
          Array [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
          ],
        ],
        "u": 1,
      },
    }
  }
  state={
    Atom {
      "_activating": false,
      "_active": true,
      "_alive": true,
      "_currentEvent": Object {
        "type": "value",
        "value": Object {
          "canvas": Object {
            "scale": 16,
            "size": Array [
              4,
              4,
            ],
          },
          "color": Object {
            "currentColor": 0,
            "currentPalette": 0,
            "palettes": Array [],
          },
          "size": Array [
            16,
            16,
          ],
        },
      },
      "_dispatcher": Dispatcher {
        "_inLoop": 0,
        "_items": Array [
          Object {
            "fn": [Function],
            "type": "any",
          },
          Object {
            "fn": [Function],
            "type": "any",
          },
          Object {
            "fn": [Function],
            "type": "any",
          },
          Object {
            "fn": [Function],
            "type": "any",
          },
          Object {
            "fn": [Function],
            "type": "any",
          },
          Object {
            "fn": [Function],
            "type": "any",
          },
        ],
        "_removedItems": null,
        "_spies": Array [],
      },
      "_logHandlers": null,
      "_spyHandlers": null,
    }
  }
>
  <div
    className="scene-root editor-root"
    data-scene-name="editor"
  >
    <div>
      something went wrong
    </div>
  </div>
</EditorScene>
`;
